### PR TITLE
CNN encoders refactor

### PR DESCRIPTION
**Pull request by jlibovicky**
Refactor of ImageNetEncoder, SequenceCNNEncoder and SentenceCNNEncoder using the `@tensor` annotation.

### CODE REVIEW

File: neuralmonkey/encoders/imagenet_encoder.py [hunk 1]
```diff
@@ -97,56 +98,70 @@ def __init__(self,
...
-    # pylint: enable=too-many-arguments,too-many-locals
+            _, self._end_points = net_function(self.input_image)
+
+    @tensor
+    def input_image(self) -> tf.Tensor:
+        return tf.placeholder(
+            tf.float32, [None, self.HEIGHT, self.WIDTH, 3])
+
+    @tensor
+    def cnn_states(self) -> tf.Tensor:
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/imagenet_encoder.py]: jestli to někdy vrací None, tak by to mělo bejt nějak zohledněný v tý anotaci, ne?

File: neuralmonkey/encoders/imagenet_encoder.py [hunk 2]
```diff
@@ -97,56 +98,70 @@ def __init__(self,
...
+        if self._attention_layer is None:
+            return None
+
+        if self._attention_layer not in self._end_points:
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/imagenet_encoder.py]: kde se vemou tyhle end points?

File: neuralmonkey/encoders/imagenet_encoder.py [hunk 3]
```diff
@@ -97,56 +98,70 @@ def __init__(self,
...
+            raise ValueError(
+                "Network '{}' does not contain endpoint '{}'".format(
+                    self._network_type, self._attention_layer))
+
+        net_output = self._end_points[self._attention_layer]
+
+        if len(net_output.get_shape()) != 4:
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/imagenet_encoder.py]: přidal bych tenhle check už do konstruktoru.

File: neuralmonkey/encoders/imagenet_encoder.py [hunk 4]
```diff
@@ -97,56 +98,70 @@ def __init__(self,
...
+        # pylint: enable=no-member
+        return tf.reshape(
+            self.cnn_states, [-1, shape[0] * shape[1], shape[2]])
+
+    @tensor
+    def encoded(self) -> tf.Tensor:
+        if self._encoded_layer is None:
+            return tf.reduce_mean(self.cnn_states, [1, 2])
+
+        if self._encoded_layer not in self._end_points:
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/imagenet_encoder.py]: zase - kde se vemou end points a jestli to neni tenzor tak bych to dal do konstruktoru

File: neuralmonkey/encoders/imagenet_encoder.py [hunk 5]
```diff
@@ -97,56 +98,70 @@ def __init__(self,
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/imagenet_encoder.py]: ty checky většinou bejvaj až pod přiřazenim argumetnů do selfu.

File: neuralmonkey/encoders/sentence_cnn_encoder.py [hunk 6]
```diff
@@ -109,93 +110,123 @@ def __init__(self,
...
-                self.__attention_tensor = dropout(
-                    self.hidden_states, self.dropout_keep_prob,
-                    self.train_mode)
+    @tensor
+    def sentence_lengths(self) -> tf.Tensor:
+        return tf.to_int32(tf.reduce_sum(self.input_mask, 1))
 
-            self.encoded = tf.concat(encoded_tup, 1)
+    @tensor
+    def embedded_inputs(self) -> tf.Tensor:
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/sentence_cnn_encoder.py]: chceme rovnou embedded inputs, nebo embedding matrix a pak embedded inputs?

File: neuralmonkey/encoders/sentence_cnn_encoder.py [hunk 7]
```diff
@@ -109,93 +110,123 @@ def __init__(self,
...
+        for i in range(self.highway_depth):
+            highway_layer = highway(
+                highway_layer,
+                scope=("highway_layer_%s" % i))
+        highway_layer = tf.reshape(
+            highway_layer,
+            [batch_size, -1, cnn_out_size])
+
+    @tensor
+    def bidirectional_rnn(self) -> tf.Tensor:
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/sentence_cnn_encoder.py]: tohle nevrací tensor ale tupl tuplu a tuplu

File: neuralmonkey/encoders/sentence_cnn_encoder.py [hunk 8]
```diff
@@ -109,93 +110,123 @@ def __init__(self,
...
+        # pylint: enable=unsubscriptable-object
+
+    @tensor
+    def encoded(self) -> tf.Tensor:
+        # pylint: disable=unsubscriptable-object
+        return tf.concat(self.bidirectional_rnn[1], 1)
+        # pylint: enable=unsubscriptable-object
+
+    @tensor
     def _attention_tensor(self):
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/sentence_cnn_encoder.py]: tohle vvrací tensor a ta funkce pod touhle taky

File: neuralmonkey/encoders/sentence_cnn_encoder.py [hunk 9]
```diff
@@ -210,41 +241,6 @@ def _attention_mask(self):
     def vocabulary_size(self):
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/sentence_cnn_encoder.py]: vrací int

File: neuralmonkey/encoders/sequence_cnn_encoder.py [hunk 10]
```diff
@@ -46,56 +49,70 @@ def __init__(self,
...
-                    conv_relu = tf.nn.relu(tf.nn.bias_add(conv, b_filter))
-
-                    # Max-pooling over the outputs
-                    pooled = tf.reduce_max(conv_relu, 1)
-                    pooled_outputs.append(pooled)
-
-            # Combine all the pooled features
-            self.encoded = tf.concat(pooled_outputs, axis=1)
-    # pylint: enable=too-many-arguments,too-many-locals
+        self._embedding_size = embedding_size
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/sequence_cnn_encoder.py]: podtržitka teda

File: neuralmonkey/encoders/sequence_cnn_encoder.py [hunk 11]
```diff
@@ -46,56 +49,70 @@ def __init__(self,
...
+                              name="train_mode")
+
+    @tensor
+    def inputs(self) -> tf.Tensor:
+        return tf.placeholder(tf.int32,
+                              shape=[None, None],
+                              name="encoder_input")
+
+    @tensor
+    def _input_mask(self) -> tf.Tensor:
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/sequence_cnn_encoder.py]: jinde je input mask public

File: neuralmonkey/encoders/sequence_cnn_encoder.py [hunk 12]
```diff
@@ -46,56 +49,70 @@ def __init__(self,
...
+        return tf.placeholder(tf.float32, shape=[None, None],
+                              name="encoder_padding")
+    # pylint: enable=no-self-use
+
+    @tensor
+    def _embedded_inputs(self) -> tf.Tensor:
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/sequence_cnn_encoder.py]: stejna otazka jako u minulyho enkoderu - nechcem to rozdelit na emb matrix a embedded inputs? taky bych to dal public, nebo do jiný funkce. jako její součást ale to spíš nechcem protože se to bude dělit v tom dalším refakoru

**Comment by jindrahelcl** [comment]
Dušan tu 14 dní nebude

File: neuralmonkey/encoders/sequence_cnn_encoder.py (hunk 12 above)
Comment (jlibovicky) [inline neuralmonkey/encoders/sequence_cnn_encoder.py]: Je to IMHO jedno, protože tyhlety enkodéři stejně budeme rozdělovat na dvě půlky.

Comment (jindrahelcl) [inline neuralmonkey/encoders/sequence_cnn_encoder.py]: ale až za tim embedováním

File: neuralmonkey/encoders/imagenet_encoder.py (hunk 5 above)
Comment (jlibovicky) [inline neuralmonkey/encoders/imagenet_encoder.py]: Víc by mi to dávalo smysl předtím, ale je to asi jedno.

Comment (jlibovicky) [inline neuralmonkey/encoders/imagenet_encoder.py]: Že jako nejdřív zkontroluju, jestli to stojí za to a pak to začnu sypat do toho selfu.

Comment (jindrahelcl) [inline neuralmonkey/encoders/imagenet_encoder.py]: když budou po tom, tak se budou moct dát do funkce, která bere jen `self` jako parametr...

Comment (jindrahelcl) [inline neuralmonkey/encoders/imagenet_encoder.py]: když budou po tom, tak se budou moct dát do funkce, která bere jen `self` jako parametr...

File: neuralmonkey/encoders/imagenet_encoder.py (hunk 4 above)
Comment (jlibovicky) [inline neuralmonkey/encoders/imagenet_encoder.py]: je to slovník tenzorů a je v konstruktoru

File: neuralmonkey/encoders/imagenet_encoder.py (hunk 3 above)
Comment (jindrahelcl) [inline neuralmonkey/encoders/imagenet_encoder.py]: dyť to máš v konstruktoru ne? i end points i attention layer..

File: neuralmonkey/encoders/imagenet_encoder.py (hunk 4 above)
Comment (jindrahelcl) [inline neuralmonkey/encoders/imagenet_encoder.py]: šlo mi o ten check.

File: neuralmonkey/encoders/sequence_cnn_encoder.py (hunk 12 above)
Comment (jlibovicky) [inline neuralmonkey/encoders/sequence_cnn_encoder.py]: No embedování bude speciální třída, takže je jedno, jak se to udělá tady, protože odsud se to stejně smaže.

Comment (jindrahelcl) [inline neuralmonkey/encoders/sequence_cnn_encoder.py]: ok
